[TS-197] Yaga. Migration library

im-kulikov opened the pull request:
- migrator + tests
- update cli-package
- move testdb to helpers

im-kulikov committed:
[TS-197] Yaga. Migration library

- migrator + tests
- update cli-package
- move testdb to helpers

im-kulikov committed:
- cleanup
- fixes

im-kulikov committed:
remove debug message

im-kulikov committed:
- fix bad migration files naming
- add doc comments
- fix readme

im-kulikov committed:
- try to cache containers

im-kulikov committed:
- add migration cmd
- rewrite migration cmd in CircleCI

im-kulikov committed:
- ignore non sql files

im-kulikov committed:
- make empty "1512662685_sample.down.sql" file

im-kulikov committed:
- added new functionality to create new migration

im-kulikov committed:
- fix stepped migration

jenchik commented on migrate/helpers.go:
```diff
@@ -0,0 +1,174 @@
+
+	if dir, err = os.Open(path); os.IsNotExist(err) {
+		return nil, ErrDirNotExists
+	}
```
no ```dir.Close()```

im-kulikov replied:
fixed

jenchik commented on migrate/helpers.go:
```diff
@@ -0,0 +1,174 @@
+	)
+
+	for _, item := range items {
+		if f, err = os.Create(path + "/" + fmt.Sprintf(fileNameTpl, dt, name, item)); err != nil {
```
mb
```
if f, err = os.Create(fmt.Sprintf(filePathTpl, path, dt, name, item)); err != nil {
```

m1ome replied:
`path.Join` guys, it should be cross-os IMO.

im-kulikov replied:
yep..

im-kulikov replied:
fixed

jenchik commented on migrate/helpers.go:
```diff
@@ -0,0 +1,174 @@
+}
+
+// extractAttributes, such as  version, name and migration-type
+func extractAttributes(filename string) (version int64, name, mType string, err error) {
```
mb one regex?

im-kulikov replied:
I don't think so

jenchik replied:
did a lot of unnecessary work instead of executing one command

im-kulikov replied:
I didn't like regexp

jenchik replied:
very bad

jenchik commented on migrate/cmd/main.go:
```diff
@@ -0,0 +1,90 @@
+	opts = new(Options)
+)
+
+func main() {
```
mb added DSN parameter?

jenchik replied:
sorry, but we can not use it

im-kulikov replied:
yep

jenchik commented:
mb added command ```migrate:list``` ?

im-kulikov committed:
- close dir (thanks @jenchik)

im-kulikov committed:
- migration list + tests (example):

```
2018-02-19T19:54:43.948+0300    INFO    migrate/helpers.go:113  Prepare migration file: 1512662685_second.up.sql
2018-02-19T19:54:43.948+0300    INFO    migrate/helpers.go:113  Prepare migration file: 1512662684_first.up.sql
2018-02-19T19:54:43.948+0300    INFO    migrate/helpers.go:113  Prepare migration file: 1512662684_first.down.sql
2018-02-19T19:54:43.948+0300    INFO    migrate/helpers.go:113  Prepare migration file: 1512662685_second.down.sql
2018-02-19T19:54:43.948+0300    INFO    migrate/helpers.go:146  Prepare migration: first
2018-02-19T19:54:43.948+0300    INFO    migrate/helpers.go:146  Prepare migration: second
2018-02-19T19:54:43.998+0300    INFO    cmd/main.go:93  1512662684 first -> 2018-02-19 16:31:11.322626 +0000 +0000
2018-02-19T19:54:43.998+0300    INFO    cmd/main.go:93  1512662685 second -> 2018-02-19 16:31:11.333213 +0000 +0000
2018-02-19T19:54:43.998+0300    INFO    cmd/main.go:99  all done
```

im-kulikov committed:
fix tests

im-kulikov committed:
fix tests

im-kulikov committed:
review comment fixes

m1ome commented on helpers/testdb/main.go:
```diff
@@ -24,7 +25,7 @@ type Database struct {
 	if database == nil {
 		err := godotenv.Load()
-		if err != nil {
+		if err != nil && !strings.Contains(err.Error(), "no such file or directory") {
```
`const` ?

im-kulikov replied:
there's any sense?

im-kulikov replied:
it's used just to ignore message `.env no such file or directory`

m1ome commented on migrate/cmd/main.go:
```diff
@@ -0,0 +1,100 @@
+	return len(o.Addr) > 0 &&
+		len(o.User) > 0 &&
+		len(o.Database) > 0 &&
+		len(o.Password) > 0 &&
```
empty password?

m1ome commented on migrate/cmd/main.go:
```diff
@@ -0,0 +1,100 @@
+		User:     o.User,
+		Database: o.Database,
+		Password: o.Password,
+		PoolSize: 2,
```
Maybe use default?

m1ome replied:
This looks like a magic number

im-kulikov replied:
ok

m1ome commented on migrate/cmd/main.go:
```diff
@@ -0,0 +1,100 @@
+
+func main() {
+	flag.StringVar(&opts.Password, "p", "", "password")
+	flag.StringVar(&opts.Addr, "a", "", "address")
```
`localhost` is default?

im-kulikov replied:
hm... it must be `localhost:5432`

m1ome commented on migrate/cmd/main.go:
```diff
@@ -0,0 +1,100 @@
+	flag.StringVar(&opts.Addr, "a", "", "address")
+	flag.StringVar(&opts.Database, "d", "", "db name")
+	flag.StringVar(&opts.User, "u", "", "username")
+	flag.StringVar(&opts.Path, "src", "", "path to migrations")
```
by default `./migrations` ?

jenchik replied:
better by default than nothing

im-kulikov committed:
add `migrate:list` command to yaga/cli

m1ome commented on migrate/helpers.go:
```diff
@@ -0,0 +1,178 @@
+
+		log.Infof("Prepare migration file: %s", file.Name())
+
+		if data, err = ioutil.ReadFile(path + "/" + file.Name()); err != nil {
```
path.Join

im-kulikov committed:
path.Join...

im-kulikov committed:
review comments

m1ome commented on migrate/vars.go:
```diff
@@ -0,0 +1,43 @@
+	sqlCreateTable   = `
+CREATE TABLE IF NOT EXISTS ? (
+	id serial,
+	version bigint UNIQUE,
```
maybe we should add full migration name?

im-kulikov replied:
fixed

pcherednichenko reviewed (dismissed):
`dir.Close()` it is necessary to correct, and looks like everything is ok

______
_Added later:_
aaa, I see, okay

im-kulikov committed:
- migration plan / version / real-name of migration in db
